File: 백준/Gold/1167. 트리의 지름/트리의 지름.java
import java.io.*;
import java.util.*;

class Main {
    static List<Node>[] tree;
    static boolean[] vis;
    static int maxWeight = 0;
    static int x = 1;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int v = Integer.parseInt(br.readLine());

        tree = new List[v + 1];
        vis = new boolean[v + 1];

        for (int i = 0; i < v; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            int s = Integer.parseInt(st.nextToken());
            tree[s] = new ArrayList<>();

            int node = Integer.parseInt(st.nextToken());

            while (node > -1) {
                int weight = Integer.parseInt(st.nextToken());

                tree[s].add(new Node(node, weight));

                node = Integer.parseInt(st.nextToken());
            }
        }
        br.close();

        dfs(1, 0);

        Arrays.fill(vis, false);

        dfs(x, 0);

        System.out.println(maxWeight);
    }

    static void dfs(int s, int weight) {
        vis[s] = true;

        if (weight > maxWeight) {
            maxWeight = weight;
            x = s;
        }

        for (Node node : tree[s]) {
            if (!vis[node.node]) {
                dfs(node.node, weight + node.weight);
            }
        }
    }

    static class Node {
        int node;
        int weight;

        public Node(int node, int weight) {
            this.node = node;
            this.weight = weight;
        }
    }
}

File: 백준/Gold/1167. 트리의 지름/트리의 지름.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val edges = Array(n) { arrayListOf<Node>() }

    repeat(n) {
        val st = StringTokenizer(readLine())
        val u = st.nextToken().toInt() - 1
        var v = st.nextToken().toInt() - 1

        while (v >= 0) {
            edges[u].add(Node(v, st.nextToken().toInt()))
            v = st.nextToken().toInt() - 1
        }
    }
    close()

    val visit = BooleanArray(n)
    var max = 0
    var maxIndex = 0

    fun travel(u: Int, d: Int) {
        visit[u] = true

        if (d > max) {
            max = d
            maxIndex = u
        }

        for (v in edges[u]) {
            if (!visit[v.v]) {
                travel(v.v, d + v.c)
            }
        }
    }

    travel(0, 0)

    visit.fill(false)

    travel(maxIndex, 0)

    with(System.out.bufferedWriter()) {
        append("$max").flush()
        close()
    }
}

class Node(val v: Int, val c: Int)